### ISSUE TITLE
`Agent-groups send full` task not running all of the times that it is requested

### ISSUE THREAD

**Issue by Selutario** [wazuh/wazuh#15958]
|Wazuh version|Component|
|---|---|
| 4.4.0 | Wazuh cluster |

## Description
During the workload testing of 4.4-beta1 (https://github.com/wazuh/wazuh/issues/15938#issuecomment-1385106895), the following problem has been found. It turns out that manager_3 has been showing logs like these (they are seen up to 64 more times):
```
2023/01/16 13:57:13 DEBUG: [Worker CLUSTER-Workload_benchmarks_metrics_B235_manager_3] [Agent-groups recv] The master's checksum and the worker's checksum are different. Local checksum: 394e44a39235255b70848c1a39abe2e5161edad9 | Master checksum: fa83a177d6dc7836b77d40012d3913f9d2d9cc7b.
2023/01/16 13:57:13 INFO: [Worker CLUSTER-Workload_benchmarks_metrics_B235_manager_3] [Agent-groups recv] Sent request to obtain all agent-groups information from the master node.
2023/01/16 13:57:13 INFO: [Worker CLUSTER-Workload_benchmarks_metrics_B235_manager_3] [Agent-groups recv] Finished in 0.074s. Updated 1 chunks.
```

However, in the master it is only seen 6 times in total:
```
2023/01/16 14:07:25 DEBUG: [Worker CLUSTER-Workload_benchmarks_metrics_B235_manager_3] [Agent-groups send full] Obtained 26 chunks of data in 1.156s.
2023/01/16 14:07:25 INFO: [Worker CLUSTER-Workload_benchmarks_metrics_B235_manager_3] [Agent-groups send full] Requested entire agent-groups information by the worker node. Starting.
2023/01/16 14:07:25 INFO: [Worker CLUSTER-Workload_benchmarks_metrics_B235_manager_3] [Agent-groups send full] Sending all agent-groups information from the master node database.
2023/01/16 14:07:25 DEBUG: [Worker CLUSTER-Workload_benchmarks_metrics_B235_manager_3] [Agent-groups send full] Sending chunks.
2023/01/16 14:07:27 DEBUG: [Worker CLUSTER-Workload_benchmarks_metrics_B235_manager_3] [Main] Command received: b'syn_wgc_e'
2023/01/16 14:07:27 INFO: [Worker CLUSTER-Workload_benchmarks_metrics_B235_manager_3] [Agent-groups send full] Finished in 3.118s. Updated 26 chunks.
```

## Checks

The following elements have been updated or reviewed (should also be checked if no modification is required):
- [x] Tests (unit tests, API integration tests).
- [x] Changelog.
- [x] Documentation.
- [x] Integration test mapping (using `api/test/integration/mapping/_test_mapping.py`).

**Comment by Selutario** [comment]
## Status update
Although this problem seems to occur for all workers (in the master), we will focus on worker3. If we count how many times the worker has requested the information of all the agent-groups from the master, we see that the number amounts to 80:
```
$ grep "manager_3\] \[Main\] Command received: b'syn_w_g_c'" cluster.log | wc -l
80
```

However, the task that should be executed when receiving said command only appears 6 times:
```
$ grep "manager_3\] \[Agent-groups send full\] Requested entire agent-groups" cluster.log | wc -l
6
```

When the master receives the `syn_w_g_c` command from a worker, it creates a task to run a function:
https://github.com/wazuh/wazuh/blob/90134ed0cc7a4479a216af6d10743bd013fe5700/framework/wazuh/core/cluster/common.py#L191-L195

The set_up_coro() function is a reference to this:
https://github.com/wazuh/wazuh/blob/90134ed0cc7a4479a216af6d10743bd013fe5700/framework/wazuh/core/cluster/master.py#L198-L200
https://github.com/wazuh/wazuh/blob/90134ed0cc7a4479a216af6d10743bd013fe5700/framework/wazuh/core/cluster/master.py#L682-L706

The above method has a problem: unlike other tasks, this doesn't print any logs right when it is executed (`[Agent-groups send full] Starting.`). However, its behavior is inherited from the same class used to receive Agent-info data. If we look for how many times the master has received the command to start the agent-info task:
```
$ grep "manager_3\] \[Main\] Command received: b'syn_a_w_m'" cluster.log | wc -l
65
```

We see that it has been executed the same number of times:
```
$ grep "manager_3\] \[Agent-info sync\] Starting." cluster.log | wc -l
65
```

It means that probably there is not any problem preventing the `send_entire_agent_groups_information` method from being executed. 

Another possible cause of the error is that, since wazuh-db is being called to get group data (shown below), if the response is not being received the function could end up blocked in multiple occasions:
https://github.com/wazuh/wazuh/blob/90134ed0cc7a4479a216af6d10743bd013fe5700/framework/wazuh/core/cluster/master.py#L694

In fact, we count the number of unexecuted `Agent-groups send full` tasks:
**2016** (number of `syn_w_g_c` command received) **- 125** (number of executed `Agent-groups send full`) **= 1891**


The number is close to the amount of `Broken pipes` that can be seen in the master's `ossec.log` file:
```
$ grep "Broken pipe" ossec.log | wc -l
1828
```

I'm now trying to reproduce this problem in my local env.

**Comment by Selutario** [comment]
## Status update

I've been going through the issues of the agent-groups epic (https://github.com/wazuh/wazuh/issues/10771) to better understand the complete agent-groups synchronization workflow. 

In my small-scale tests, whenever I modify an agent's groups in a worker's db:
```
/scratches/send_msg.py 'global set-agent-groups {"mode":"override","sync_status":"synced","data":[{"id":1,"groups":["default","new_test_group"]},{"id":2,"groups":["default","new_test_group"]},{"id":3,"groups":["default"]}]}'
```

Synchronization of complete agent-groups information succeeds:
```
2023/01/23 11:11:52 INFO: [Worker worker1] [Agent-groups recv] Starting.
2023/01/23 11:11:52 DEBUG: [Worker worker1] [Agent-groups recv] 1/1 chunks updated in wazuh-db in 0.000s.
2023/01/23 11:11:52 DEBUG: [Worker worker1] [Agent-groups recv] Obtained 1 chunks of data in 0.000s.
2023/01/23 11:11:52 DEBUG: [Worker worker1] [Agent-groups recv] The master's checksum and the worker's checksum are different. Local checksum: 16b0b140532e41f5f02f5407abc3e50a6626ca21 | Master checksum: 8324cc76e7ad9a1420ff6d09083e7e7d525eb87e.
2023/01/23 11:11:52 INFO: [Worker worker1] [Agent-groups recv] Sent request to obtain all agent-groups information from the master node.
2023/01/23 11:11:52 INFO: [Worker worker1] [Agent-groups recv] Finished in 0.003s. Updated 1 chunks.
2023/01/23 11:11:52 INFO: [Worker worker1] [Agent-groups recv full] Starting.
2023/01/23 11:11:52 DEBUG: [Worker worker1] [Agent-groups recv full] 1/1 chunks updated in wazuh-db in 0.001s.
2023/01/23 11:11:52 INFO: [Worker worker1] [Agent-groups recv full] Finished in 0.003s. Updated 1 chunks.
```

During these tests, I ran into a problem whereby the global group hash remained unchanged despite modifying the groups or even the hash of each one. The reason was that `wazuh-db` uses a cache system so it is necessary to modify the group information through the command used above instead of through SQL queries.

I'm now trying to reproduce the test environment locally with 50k agents. In the meantime, I created a script to periodically get `wazuh-db` stats from all nodes via the `/cluster/{}/daemons/stats` endpoint. I'll run the test again where the problem was found along with this script. A debug message has also been included to determine if the agent deletion is actually working fine.

**Comment by Selutario** [comment]
## Status update

I have managed to reproduce the bug using an environment like this:
- 1 master node
- 3 worker nodes
- 50000 agents

I haven't tried using fewer agents yet, it may be possible too. 

The problem happens when 2 or more workers ask the master for the complete information of agent-groups. The master opens a connection to wazuh-db for each worker. However, two of them end up breaking (I'm still trying to figure out the reason). As a consequence, the master only responds to one of the workers.

Here we can see the logs of 3 workers requesting full agent-groups data at the same time:
**Worker 2:**
```
        2023/01/24 11:17:37 INFO: [Worker worker2] [Agent-groups recv] Starting.
#1 ---> 2023/01/24 11:17:37 INFO: [Worker worker2] [Agent-groups recv] Sent request to obtain all agent-groups information from the master node.
        2023/01/24 11:17:37 INFO: [Worker worker2] [Agent-groups recv] Finished in 0.046s. Updated 1 chunks.
        2023/01/24 11:17:38 INFO: [Worker worker2] [Agent-groups recv full] Starting.
        2023/01/24 11:17:40 INFO: [Worker worker2] [Agent-groups recv full] Finished in 2.584s. Updated 40 chunks.
        2023/01/24 11:17:47 INFO: [Worker worker2] [Agent-groups recv] Starting.
        2023/01/24 11:17:47 INFO: [Worker worker2] [Agent-groups recv] Finished in 0.054s. Updated 1 chunks.
        2023/01/24 11:17:57 INFO: [Worker worker2] [Agent-groups recv] Starting.
        2023/01/24 11:17:57 INFO: [Worker worker2] [Agent-groups recv] Finished in 0.053s. Updated 1 chunks.
```

**Worker 3:**
```
        2023/01/24 11:17:37 INFO: [Worker worker3] [Agent-groups recv] Starting.
#1 ---> 2023/01/24 11:17:37 INFO: [Worker worker3] [Agent-groups recv] Sent request to obtain all agent-groups information from the master node.
        2023/01/24 11:17:37 INFO: [Worker worker3] [Agent-groups recv] Finished in 0.052s. Updated 1 chunks.
        2023/01/24 11:17:47 INFO: [Worker worker3] [Agent-groups recv] Starting.
#2 ---> 2023/01/24 11:17:47 INFO: [Worker worker3] [Agent-groups recv] Sent request to obtain all agent-groups information from the master node.
        2023/01/24 11:17:47 INFO: [Worker worker3] [Agent-groups recv] Finished in 0.042s. Updated 1 chunks.
        2023/01/24 11:17:48 INFO: [Worker worker3] [Agent-groups recv full] Starting.
        2023/01/24 11:17:50 INFO: [Worker worker3] [Agent-groups recv full] Finished in 2.612s. Updated 40 chunks.
        2023/01/24 11:17:57 INFO: [Worker worker3] [Agent-groups recv] Starting.
        2023/01/24 11:17:57 INFO: [Worker worker3] [Agent-groups recv] Finished in 0.073s. Updated 1 chunks.
```

**Worker 1:**
```
        2023/01/24 11:17:37 INFO: [Worker worker1] [Agent-groups recv] Starting.
#1 ---> 2023/01/24 11:17:37 INFO: [Worker worker1] [Agent-groups recv] Sent request to obtain all agent-groups information from the master node.
        2023/01/24 11:17:37 INFO: [Worker worker1] [Agent-groups recv] Finished in 0.054s. Updated 1 chunks.
        2023/01/24 11:17:47 INFO: [Worker worker1] [Agent-groups recv] Starting.
#2 ---> 2023/01/24 11:17:47 INFO: [Worker worker1] [Agent-groups recv] Sent request to obtain all agent-groups information from the master node.
        2023/01/24 11:17:47 INFO: [Worker worker1] [Agent-groups recv] Finished in 0.046s. Updated 1 chunks.
        2023/01/24 11:17:57 INFO: [Worker worker1] [Agent-groups recv] Starting.
#3 ---> 2023/01/24 11:17:57 INFO: [Worker worker1] [Agent-groups recv] Sent request to obtain all agent-groups information from the master node.
        2023/01/24 11:17:57 INFO: [Worker worker1] [Agent-groups recv] Finished in 0.053s. Updated 1 chunks.
        2023/01/24 11:17:58 INFO: [Worker worker1] [Agent-groups recv full] Starting.
        2023/01/24 11:18:00 INFO: [Worker worker1] [Agent-groups recv full] Finished in 2.637s. Updated 40 chunks.
```

The request `#1` to obtain all agent-groups information is performed at `11:17:37` in the 3 nodes, but a response (`Agent-groups recv full`) is only received in worker2. Then, both worker3 and worker1 make a second (`#2`) request, although master only responds to worker2. Finally, the worker3 is the only one requesting the information for the third (`#3`) time, so now it works. In the master's `ossec.log` this could be seen:
```
2023/01/24 11:17:37 wazuh-db: ERROR: at run_worker(): OS_SendSecureTCP(35): Broken pipe (32)
2023/01/24 11:17:37 wazuh-db: ERROR: at run_worker(): OS_SendSecureTCP(36): Broken pipe (32)
2023/01/24 11:17:47 wazuh-db: ERROR: at run_worker(): OS_SendSecureTCP(35): Broken pipe (32)
```

## How to reproduce
1. I registered 50k agents (`never_connected`).
2. I created a group named `new_test_group` and assigned it to all agents.
3. I send a message to the wazuh-db service of the three worker nodes at the same time. This way, the groups assigned to some of the agents are modified so the master's and worker's checksums do not match:
    ```
    # /scratches/send_msg.py 'global set-agent-groups {"mode":"override","sync_status":"synced","data":[{"id":1,"groups":["default","new_test_group"]},{"id":2,"groups":["default","new_test_group"]},{"id":3,"groups":["default"]}]}
    ```

**Comment by Selutario** [comment]
## Status update

This error was showing up in the master node everytime it tried to communicate with wazuh-db in multiple concurrent tasks (`SendStringTask` for worker1, worker2 and worker3 in this case):
```
2023/01/25 11:29:14 INFO: [Master] [Local agent-groups] Finished in 0.022s.
2023/01/25 11:29:14 INFO: [Worker worker1] [Agent-groups send] Starting.
2023/01/25 11:29:14 INFO: [Worker worker2] [Agent-groups send] Starting.
2023/01/25 11:29:14 INFO: [Worker worker1] [Agent-groups send] Finished in 0.019s. Updated 1 chunks.
2023/01/25 11:29:14 INFO: [Worker worker2] [Agent-groups send] Finished in 0.018s. Updated 1 chunks.
2023/01/25 11:29:14 INFO: [Worker worker3] [Agent-groups send] Starting.
2023/01/25 11:29:14 INFO: [Worker worker3] [Agent-groups send] Finished in 0.014s. Updated 1 chunks.
Unhandled error in exception handler
context: {'message': 'Task was destroyed but it is pending!', 'task': <Task pending name='Task-31' coro=<MasterHandler.send_entire_agent_groups_information() done, defined at /var/ossec/framework/python/lib/python3.9/site-packages/wazuh-4.4.0-py3.9.egg/wazuh/core/cluster/master.py:682> wait_for=<Future pending cb=[<TaskWakeupMethWrapper object at 0x7f6eba268490>()]> cb=[SendStringTask.done_callback()]>}
Traceback (most recent call last):
  File "/var/ossec/framework/python/lib/python3.9/asyncio/base_events.py", line 1775, in call_exception_handler
    self._exception_handler(self, context)
  File "/var/ossec/framework/python/lib/python3.9/site-packages/wazuh-4.4.0-py3.9.egg/wazuh/core/cluster/common.py", line 1579, in asyncio_exception_handler
    logging.error(f"Unhandled exception: {context['exception']} {context['message']}\n"
KeyError: 'exception'
Unhandled error in exception handler
context: {'message': 'Task was destroyed but it is pending!', 'task': <Task pending name='Task-32' coro=<MasterHandler.send_entire_agent_groups_information() done, defined at /var/ossec/framework/python/lib/python3.9/site-packages/wazuh-4.4.0-py3.9.egg/wazuh/core/cluster/master.py:682> wait_for=<Future pending cb=[<TaskWakeupMethWrapper object at 0x7f6eba268460>()]> cb=[SendStringTask.done_callback()]>}
Traceback (most recent call last):
  File "/var/ossec/framework/python/lib/python3.9/asyncio/base_events.py", line 1775, in call_exception_handler
    self._exception_handler(self, context)
  File "/var/ossec/framework/python/lib/python3.9/site-packages/wazuh-4.4.0-py3.9.egg/wazuh/core/cluster/common.py", line 1579, in asyncio_exception_handler
    logging.error(f"Unhandled exception: {context['exception']} {context['message']}\n"
KeyError: 'exception'
2023/01/25 11:29:15 INFO: [Worker worker2] [Agent-groups send full] Requested entire agent-groups information by the worker node. Starting.
2023/01/25 11:29:15 INFO: [Worker worker2] [Agent-groups send full] Sending all agent-groups information from the master node database.
2023/01/25 11:29:17 INFO: [Master] [Local integrity] Starting.
2023/01/25 11:29:17 INFO: [Master] [Local integrity] Finished in 0.003s. Calculated metadata of 39 files.
```

Only one task could finish, the rest were closed after sending a message to wdb but before getting its response. The problem seems to be a [bug](https://bugs.python.org/issue46309) in the [Streams](https://docs.python.org/3/library/asyncio-stream.html) library. This library is used in the cluster to communicate with the wazuh-db socket, allowing other tasks to be performed asynchronously until the response is ready:

> No reference was kept to a task created by StreamReaderProtocol and was therefore eligible for garbage collection under certain circumstances.

A solution for this problem is implemented in Python 3.11 and 3.10.8, but since Wazuh 4.4 will still use Python 3.9.9, it is necessary to include a workaround to store a reference of each task in which Streams is used:
https://github.com/wazuh/wazuh/blob/90134ed0cc7a4479a216af6d10743bd013fe5700/framework/wazuh/core/cluster/common.py#L129-L148

Now the reference is kept until `self.done_callback` is executed. At that point it is safe to allow the garbage collector to kill the task. By the time it is deleted, the master has already run the `Agent-groups send full` task for each of the workers:
```
2023/01/25 13:58:03 INFO: [Master] [Local agent-groups] Finished in 0.021s.
2023/01/25 13:58:03 INFO: [Worker worker2] [Agent-groups send] Starting.
2023/01/25 13:58:03 INFO: [Worker worker1] [Agent-groups send] Starting.
2023/01/25 13:58:03 INFO: [Worker worker3] [Agent-groups send] Starting.
2023/01/25 13:58:03 INFO: [Worker worker1] [Agent-groups send] Finished in 0.016s. Updated 1 chunks.
2023/01/25 13:58:03 INFO: [Worker worker2] [Agent-groups send] Finished in 0.019s. Updated 1 chunks.
2023/01/25 13:58:04 INFO: [Worker worker3] [Agent-groups send] Finished in 0.016s. Updated 1 chunks.
2023/01/25 13:58:05 INFO: [Worker worker3] [Agent-groups send full] Requested entire agent-groups information by the worker node. Starting.
2023/01/25 13:58:05 INFO: [Worker worker3] [Agent-groups send full] Sending all agent-groups information from the master node database.
2023/01/25 13:58:05 INFO: [Worker worker1] [Agent-groups send full] Requested entire agent-groups information by the worker node. Starting.
2023/01/25 13:58:05 INFO: [Worker worker1] [Agent-groups send full] Sending all agent-groups information from the master node database.
2023/01/25 13:58:05 INFO: [Worker worker2] [Agent-groups send full] Requested entire agent-groups information by the worker node. Starting.
2023/01/25 13:58:05 INFO: [Worker worker2] [Agent-groups send full] Sending all agent-groups information from the master node database.
2023/01/25 13:58:06 INFO: [Master] [Local integrity] Starting.
2023/01/25 13:58:06 INFO: [Master] [Local integrity] Finished in 0.004s. Calculated metadata of 39 files.
```

**Comment by davidjiglesias** [comment]
To be taken into account after #13265 is completed.
